xarray/core/ops.py

```diff
@@ -279,11 +279,18 @@ def count(data, axis=None):
     return sum(~isnull(data), axis=axis)
 
 
-def fillna(data, other):
+def fillna(data, other, join="left"):
     """Fill missing values in this object with data from the other object.
     Follows normal broadcasting and alignment rules.
     """
-    return where(isnull(data), other, data)
+    from .computation import apply_ufunc
+
+    def _fillna(data, other):
+        left, right = np.broadcast_arrays(data, other)
+        result = left.copy()  # view must be copied before being written
+        result[isnull(result)] = right[isnull(result)]
+        return result
+    return apply_ufunc(_fillna, data, other, join=join)
```

Comment:
It might be helpful to separate the `join` argument to `apply_ufunc` into two parts:

1. The `join` used for [aligning indexes](https://github.com/pydata/xarray/blob/4d36fe2e8a1f5504d776c2d69895bc4d5a9cd31a/xarray/core/computation.py#L340).
2. The `join` used for [aligning data variables between datasets](https://github.com/pydata/xarray/blob/4d36fe2e8a1f5504d776c2d69895bc4d5a9cd31a/xarray/core/computation.py#L347).

For `fillna`, I think we would always want `data_vars_join='left'`, though I guess it's equivalent to `data_vars_join='left'` when we constrain `value` not to contain new keys.